tests/src/python/test_qgsvectorlayerutils.py

```diff
@@ -258,23 +258,26 @@ def testCreateFeature(self):
         # layer with default value expression
         layer.setDefaultValueDefinition(2, QgsDefaultValue('3*4'))
         f = QgsVectorLayerUtils.createFeature(layer)
-        self.assertEqual(f.attributes(), [NULL, NULL, 12.0])
-        # we expect the default value expression to take precedence over the attribute map
+        self.assertEqual(f.attributes(), [NULL, NULL, 12])
+        # we do not expect the default value expression to take precedence over the attribute map
         f = QgsVectorLayerUtils.createFeature(layer, attributes={0: 'a', 2: 6.0})
-        self.assertEqual(f.attributes(), ['a', NULL, 12.0])
+        self.assertEqual(f.attributes(), ['a', NULL, 6.0])
         # layer with default value expression based on geometry
         layer.setDefaultValueDefinition(2, QgsDefaultValue('3*$x'))
         f = QgsVectorLayerUtils.createFeature(layer, g)
+        #adjusted so that input value and output feature are the same
         self.assertEqual(f.attributes(), [NULL, NULL, 300.0])
         layer.setDefaultValueDefinition(2, QgsDefaultValue(None))
 
         # test with violated unique constraints
         layer.setFieldConstraint(1, QgsFieldConstraints.ConstraintUnique)
         f = QgsVectorLayerUtils.createFeature(layer, attributes={0: 'test_1', 1: 123})
-        self.assertEqual(f.attributes(), ['test_1', 128, NULL])
+        # since field 1 has Unique Constraint, it ignores value 123 that already has been set
+        self.assertEqual(f.attributes(), ['test_1', NULL, NULL])
         layer.setFieldConstraint(0, QgsFieldConstraints.ConstraintUnique)
+        # since field 0 and 1 already have values test_1 and 123, the output must be NULL
         f = QgsVectorLayerUtils.createFeature(layer, attributes={0: 'test_1', 1: 123})
-        self.assertEqual(f.attributes(), ['test_4', 128, NULL])
+        self.assertEqual(f.attributes(), [NULL, NULL, NULL])
```

Comment:
Not sure about that one. The output should be a new unique value instead of the conflicting one provided in attributes, no?

Reply:
In this particular data source there is no mechanism to auto increment a value, so the output is all NULLs.
I agree with you that it should happen when we are using data sources like postgresql.

When I ran this test locally, the only way of getting this check with this data source was to compare with [NULL, NULL, NULL]

Reply:
If we do not have yet a test with a PG layer with a sequence to get the new unique value from the provider it would be really nice to have it, so we could check if the value is correctly calculated.

Reply:
Interesting.
The current state (i.e. before this pull request) of the test suggests that there is a mechanism in place (attributes: `test_1` result: `test_4`). Or am I misinterpreting something here?

Reply:
I must confess I was confused as well, but when I ran the tests locally I could see that it did not output as the code suggested. I agree with @elpaso , I'll study other QGIS' test code to add a createFeatures test from PG.

Reply:
It comes from https://github.com/qgis/QGIS/blob/master/src/core/qgsvectorlayerutils.cpp#L214

Reply:
... and https://github.com/qgis/QGIS/blob/master/src/core/qgsvectorlayerutils.cpp#L234

Reply:
Now I'm confused, because running locally it was ignoring the case you are mentioning.

Maybe I should test again on c++

Reply:
@m-kuhn and @elpaso I ran test case that we have doubts with the c++ debugger turned on and I have realised that the v=uniqueValue never is called on the last `if` (bellow)
```
if ( checkUnique && fields.at( idx ).constraints().constraints() & QgsFieldConstraints::ConstraintUnique )
    {
      if ( QgsVectorLayerUtils::valueExists( layer, idx, v ) )
      {
        // unique constraint violated
        QVariant uniqueValue = QgsVectorLayerUtils::createUniqueValue( layer, idx, v );
        if ( uniqueValue.isValid() )
          v = uniqueValue;
      }
    }
```
To ensure this behaviour of this test (input: 'test_1' output: 'test_4') we must change the implementation. Maybe we should put an extra check inside the first `if` in order to create the unique value. I'm still working on that, if any of you guys have any suggestion, please feel free to share =]

Reply:
Changed the first if to:
```
if ( attributes.contains( idx ) )
    {
      v = attributes.value( idx );
      if ( fields.at( idx ).constraints().constraints() & QgsFieldConstraints::ConstraintUnique
           && QgsVectorLayerUtils::valueExists( layer, idx, v ) )
      {
        // unique constraint violated
        QVariant uniqueValue = QgsVectorLayerUtils::createUniqueValue( layer, idx, v );
        if ( uniqueValue.isValid() )
          v = uniqueValue;
      }
      checkUnique = false;
    }
```
And changed back the tests you have pointed out.